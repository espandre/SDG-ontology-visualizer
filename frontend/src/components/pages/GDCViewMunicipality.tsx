import {
  Stack,
  Select,
  Flex,
  Container,
  Text,
  Spacer,
  Button,
  Modal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalFooter,
  ModalBody,
  ModalCloseButton,
  useDisclosure,
  Tabs,
  TabList,
  TabPanels,
  Tab,
  TabPanel,
  SimpleGrid,
  Alert,
  AlertIcon,
  AlertTitle,
  AlertDescription,
  CloseButton,
  Center,
} from '@chakra-ui/react';
import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';

import { MunicipalityInfo, Municipality } from '../../types/municipalityTypes';
import { getAvailableYears } from '../../api/data';
import {
  getMunicipalityInfo,
  getAllMunicipalities,
  getSimilarMunicipalities,
} from '../../api/municipalities';

import MunicipalityInfoView from '../molecules/MunicipalityInfo';
import GDCView from '../molecules/GDCView';
import MunicipalityButton from '../atoms/MunicipalityButton';

type GDCViewParams = {
  municipality: string;
};

const ViewMunicipality: React.FC = () => {
  const { municipality } = useParams<GDCViewParams>();
  const [availableYears, setAvailableYears] = useState<Array<number>>();
  const [selectedYear, setSelectedYear] = useState<number>(-1);

  const [municipalityInfo, setMunicipalityInfo] = useState<MunicipalityInfo>();

  const [allMunicipalities, setAllMunicipalities] = useState<Municipality[]>();
  const [similarMunicipalities, setSimilarMunicipalities] = useState<Municipality[]>();

  const [showDataAlert, setShowDataAlert] = useState<boolean>(true);

  const { isOpen, onOpen, onClose } = useDisclosure();

  const loadData = async (muniCode: string) => {
    const data = await Promise.all([getAvailableYears(muniCode), getMunicipalityInfo(muniCode)]);

    const years: number[] = data[0];
    const muniInfo: MunicipalityInfo = data[1];

    setAvailableYears(years.sort());
    setSelectedYear(years[years.length - 1]);
    setMunicipalityInfo(muniInfo);

    const municipalities = await Promise.all([
      getAllMunicipalities(),
      getSimilarMunicipalities(muniCode),
    ]);
    setAllMunicipalities(municipalities[0]);
    setSimilarMunicipalities(municipalities[1]);
  };

  useEffect(() => {
    loadData(municipality);
  }, []);

  const onChangeYear = (evt: React.ChangeEvent<HTMLSelectElement>) => {
    setSelectedYear(parseFloat(evt.currentTarget.value));
  };

  const name = municipalityInfo === undefined ? '' : municipalityInfo.name;

  let dataAlert = null;
  if (showDataAlert) {
    dataAlert = (
      <Flex align="center" justify="center" justifyContent="space-evenly" m="0px" p="0px">
        <Center w="550px">
          <Alert status="error">
            <AlertIcon />
            <AlertTitle mr={2}>NOTICE!</AlertTitle>
            <AlertDescription>All data is fictional.</AlertDescription>
            <CloseButton
              position="absolute"
              right="8px"
              top="8px"
              onClick={() => setShowDataAlert(false)}
            />
          </Alert>
        </Center>
      </Flex>
    );
  }

  return (
    <Stack>
      <MunicipalityInfoView info={municipalityInfo} />
      <Flex align="center" justify="center" justifyContent="space-evenly" m="0px" p="0px">
        <Stack w={{ base: '900px', '2xl': '1450px' }} bg="white" m="0px" spacing="10">
          <Container align="right" justify="right" justifyContent="space-evenly" p="1em">
            <Flex>
              <Button onClick={onOpen}>Compare with ...</Button>
              <Spacer />
              <Stack direction="row">
                <Text size="md" p="0.4em">
                  Year:
                </Text>
                <Select value={selectedYear} onChange={onChangeYear} w="100px">
                  {availableYears &&
                    availableYears.map((year) => (
                      <option key={year} value={year}>
                        {year}
                      </option>
                    ))}
                </Select>
              </Stack>
            </Flex>
          </Container>
          {dataAlert}
          <GDCView
            key={selectedYear}
            year={selectedYear}
            municipality={name}
            municipalityCode={municipality}
          />
        </Stack>
      </Flex>
      <Modal isOpen={isOpen} onClose={onClose} size="5xl">
        <ModalOverlay />
        <ModalContent>
          <ModalHeader>Select municipality to compare with</ModalHeader>
          <ModalCloseButton />
          <ModalBody>
            <Tabs isLazy>
              <TabList>
                <Tab>Similar</Tab>
                <Tab>All</Tab>
              </TabList>
              <TabPanels>
                <TabPanel>
                  <SimpleGrid columns={3} spacing={10}>
                    {similarMunicipalities &&
                      similarMunicipalities.map((mun) => (
                        <MunicipalityButton
                          key={mun.code}
                          municipality={mun}
                          url={`/gdc/compare/${municipality}/${mun.code}`}
                        />
                      ))}
                  </SimpleGrid>
                </TabPanel>
                <TabPanel>
                  <SimpleGrid columns={3} spacing={10}>
                    {allMunicipalities &&
                      allMunicipalities.map((mun) => (
                        <MunicipalityButton
                          key={mun.code}
                          municipality={mun}
                          url={`/gdc/compare/${municipality}/${mun.code}`}
                        />
                      ))}
                  </SimpleGrid>
                </TabPanel>
              </TabPanels>
            </Tabs>
          </ModalBody>

          <ModalFooter>
            <Button colorScheme="blue" mr={3} onClick={onClose}>
              Close
            </Button>
          </ModalFooter>
        </ModalContent>
      </Modal>
    </Stack>
  );
};

export default ViewMunicipality;
