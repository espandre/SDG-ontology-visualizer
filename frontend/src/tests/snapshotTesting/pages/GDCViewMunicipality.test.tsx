import React from 'react';
import renderer from 'react-test-renderer';
import { Route, MemoryRouter } from 'react-router-dom';

import GDCViewMunicipality from '../../../components/pages/GDCViewMunicipality';

it('renders when there are no items', () => {
  const tree = renderer
    .create(
      <MemoryRouter initialEntries={['gdc/view/no.5001']}>
        <Route path="gdc/view/:municipality">
          <GDCViewMunicipality />
        </Route>
      </MemoryRouter>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
