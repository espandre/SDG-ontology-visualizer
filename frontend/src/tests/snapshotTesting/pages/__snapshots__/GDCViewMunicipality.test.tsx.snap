// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when there are no items 1`] = `
<div
  className="chakra-stack css-eao00i"
>
  <div
    className="css-ixqi0y"
    spacing="10"
  >
    <div
      className="chakra-stack css-efzghl"
    >
      <h2
        className="chakra-heading css-kbn7if"
      >
        Unknown municipality
      </h2>
    </div>
  </div>
  <div
    className="css-1s3qarl"
  >
    <div
      className="chakra-stack css-zww36l"
    >
      <div
        align="right"
        className="chakra-container css-zxqcz"
        justify="right"
      >
        <div
          className="css-k008qs"
        >
          <button
            className="chakra-button css-oqfhsh"
            onClick={[Function]}
            type="button"
          >
            Compare with ...
          </button>
          <div
            className="css-17xejub"
          />
          <div
            className="chakra-stack css-1q3uncw"
          >
            <p
              className="chakra-text css-140q4f4"
            >
              Year:
            </p>
            <div
              className="chakra-select__wrapper css-1wmuqwq"
            >
              <select
                className="chakra-select css-1p4gbel"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                value={-1}
              />
              <div
                className="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden={true}
                  className="chakra-select__icon"
                  focusable={false}
                  role="presentation"
                  style={
                    Object {
                      "color": "currentColor",
                      "height": "1em",
                      "width": "1em",
                    }
                  }
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="css-1s3qarl"
      >
        <div
          className="css-48y5oj"
        >
          <div
            className="chakra-alert css-1hwd9xd"
            role="alert"
          >
            <span
              className="chakra-alert__icon css-1c6vscd"
            >
              <svg
                className="chakra-icon css-fi0ww0"
                focusable={false}
                viewBox="0 0 24 24"
              >
                <path
                  d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <div
              className="chakra-alert__title css-1qkvt8e"
            >
              NOTICE!
            </div>
            <div
              className="chakra-alert__desc css-6n7j50"
            >
              All data is fictional.
            </div>
            <button
              aria-label="Close"
              className="css-k60dwl"
              onClick={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        className="css-172bykr"
      >
        <div
          className="chakra-stack css-eao00i"
        >
          <div
            className="chakra-spinner css-1nasj0w"
          >
            <span
              className="css-f8n5zr"
            >
              Loading...
            </span>
          </div>
          <p
            className="chakra-text css-0"
          >
            Loading...
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
